### ISSUE TITLE
[UX] Settings form for contact module

### ISSUE THREAD

**Issue by opi** [backdrop/backdrop-issues#2990]
## Describe your issue or idea

2 different ideas lead me here :  

1- One can't easily (aka without custom code) remove the "Send me a copy" checkbox in the site wide and user contact forms.
2- The contact module has settings stored in the `config/active/contact.settings.json`, which are about threshold, but those are not available through admin interface.

( maybe related meta issue https://github.com/backdrop/backdrop-issues/issues/1168 )

### Expected behavior (if reporting a bug)

The "Admin > Structure > Contact form" (`admin/structure/contact`) section should provide a global settings form, in addition to the *Categories* management form

What do you think about : 
- Having a settings form, at least to disable the "copy" checkbox on contact forms
- Allowing administrator to define threshold limit and window settings

---
PR by @opi: https://github.com/backdrop/backdrop/pull/2105
Follow-up PR by @klonos: https://github.com/backdrop/backdrop/pull/2226

**Comment by opi** [comment]
https://github.com/backdrop/backdrop/pull/2105 PR only implements the settings form with only the "show copy checkbox" item.

**Comment by opi** [comment]
Shameless ping for feedback @jenlampton @quicksketch @serundeputy :blush:

**Comment by olafgrabienski** [comment]
@opi Good idea to make the "Send yourself a copy" setting available. Is there a certain reason to implement it as a global setting instead of making it an option for each category?

**Comment by opi** [comment]
making it an option for each category requires some #states or #ajax to work properly, i guess. And I don't know if it's relevant...

**Comment by jenlampton** [comment]
> I need some help with english wording, for field title and description

Your wording for the title is Perfect!  I don't think a description is necessary at all, as the title is clear.

> The contact module has settings stored in the config/active/contact.settings.json, which are about threshold, but those are not available through admin interface.

I'm not sure an admin interface is necessary for this setting as it's a fairly advanced option and the default value should work for almost everyone. The value is saved in config so that other modules (both contrib or custom) can change it as necessary, but it's not a setting that most people will ever need to think about.

**Comment by olafgrabienski** [comment]
> an option for each category (...) I don't know if it's relevant

Not sure about the relevance because I tend to use *Webform* even for contact forms. I saw however that in the *Contact* module configuration there are some settings per category. In contrast, the new setting would be the only global setting, which seems strange at first sight. Could however make sense , e.g. because user contact forms wouldn't be covered by category settings.

**Comment by jenlampton** [comment]
We know we have a use case for a global setting, since @opi has one. We
don't know if there's a use case for per-category settings. I'd prefer not
to build anything unless we know it will benefit someone (and preferably
80% of all someone's). Plus, that change could easily be made by a
contribution module.

**Comment by opi** [comment]
> I'm not sure an admin interface is necessary for this setting as it's a fairly advanced option and the default value should work for almost everyone.

It's a personal opinion, but I don't like settings to be available only from config file.
I can create another PR with an "advanced settings" fieldset, collapsed, once this one will be merged.

There is a lot of other topic about *a better contact module in core*, like storage of submissions, having a contact form block, drag'n'drop order of categories, and other settings. Some interesting thoughts come from https://github.com/backdrop/backdrop-issues/issues/1168. Maybe we can create a meta issue ?

**Comment by jenlampton** [comment]
Yes, a metal issue is a good idea.

**Comment by opi** [comment]
UP! ;)
a [META issue about contact module improvements](https://github.com/backdrop/backdrop-issues/issues/3023) is alive.

What about this particular issue about contact module settings form, any thoughts ? reviews welcome.

**Comment by jenlampton** [comment]
> What about this particular issue about contact module settings form

I'm don't like the idea of including a settings form for a single option, but since it does look like we are about to have a bunch more options based on https://github.com/backdrop/backdrop-issues/issues/3023, starting here is a good idea IMHO 👍

**Comment by jenlampton** [comment]
Not sure this is going to make it as a new feature for 1.10, but as a [UX] issue there's a chance. Bumping milestone to 1.9.6 to see if it qualifies and can get done in the next 15 days.

**Comment by opi** [comment]
If we go https://github.com/backdrop/backdrop-issues/issues/3047 way, maybe this setting form is useless and the "personal copy" checkbox could be disabled among other fields in a nice drag'n'drop interface...

**Comment by quicksketch** [comment]
@herbdool filed a great PR in #3110 that adds flood control options to both the user login and contact forms.

I think it would be great to pull out the contract-module related portions of that PR, and include them here, so we would both: A) have good reason for making a new settings form (with more than one option on it) and B) avoid conflicts between these two PRs.

**Comment by opi** [comment]
PR https://github.com/backdrop/backdrop/pull/2105 updated with flood control settings.

Preview: (*note that flood control fieldset is collapsed by default in the PR*)
![bd_contact_form_with_flood](https://user-images.githubusercontent.com/159548/40046143-7d69d372-582c-11e8-9aaf-eb33292efcd9.png)

**Comment by herbdool** [comment]
Thanks @opi! I agree with collapsing it by default.

As per @quicksketch comment on the other thread, we should avoid using the words "flood control" in the title at least since it's technical. Maybe just *Send Limit* is enough here.

I would add some help text before the buttons since it's confusing and people might not realize it's a combination of the two fields: X emails *per* X time window. Something like:

> Configure thresholds for the contact form to minimize malicious users sending multiple emails. Control the number of emails that can be sent in a window of time by anyone with the same IP address.

**Comment by quicksketch** [comment]
Ideally you'd display these fields all together, such as:

```
Limit to [ 5 ] messages per [ 30 min | v ]
```

Where `[ 5 ]` is a number field and `[ 30 min | v]` is a dropdown select.

This can be a hassle to accomplish, as you typically need a theme function to display fields like this. But it makes for much better UX.

**Comment by herbdool** [comment]
I'm up for learning how (for the login settings). Let's see if @opi or I figure it out first!

**Comment by opi** [comment]
PR https://github.com/backdrop/backdrop/pull/2105 updated with inline fields, thanks @herbdool for the `container-inline` tip !

![bd_contact_form_with_flood_inline](https://user-images.githubusercontent.com/159548/40168184-20bfa2c6-59c3-11e8-9927-43f5e9ce143e.png)

I did not use the `#prefix` key, but proper `#title` ; I think this is way better for accessibility.

**Comment by herbdool** [comment]
@opi I used both title and prefix. And a setting for making title invisible (but visible to screenreaders). I think both are good since "Limit to" doesn't work as well as a title

**Comment by mikemccaffrey** [comment]
> I would add some help text before the buttons since it's confusing and people might not realize it's a combination of the two fields: X emails per X time window.

Yes, it kinda seems confusing to have the time window as an option. Why can't we ask the users to just set the number of messages per hour? What are the use cases where that would not be a convenient way to set the rate?

**Comment by BWPanda** [comment]
Grammar fix for the checkbox description text: add an 's' to 'form' (make it plural) and add a full-stop ('.') at the end.

**Comment by opi** [comment]
> @herbdool I used both title and prefix. And a setting for making title invisible (but visible to screenreaders). I think both are good since "Limit to" doesn't work as well as a title

By doing that, screenreaders have duplicate informations, both title and prefix : "Limit Limit to". Not sure this is what we want.

**Comment by opi** [comment]
@BWPanda thanks, fixed.

**Comment by quicksketch** [comment]
Review feedback: https://github.com/backdrop/backdrop/pull/2105#pullrequestreview-122798319

As posted there, as this is an entirely new form and page, saving this for 1.11.0 seems reasonable. Especially as we're planning on expanding this out further with more options in the near future.

**Comment by opi** [comment]
Thanks for your review @quicksketch. All your concerns were addressed.

**Comment by quicksketch** [comment]
Looks great! I've merged https://github.com/backdrop/backdrop/pull/2105 into 1.x for 1.11.0. Hopefully this will unblock some of the other Contact module related issues. Thanks @opi and everyone for the reviews!

**Comment by klonos** [comment]
Sorry for coming late to the party. A couple of comments:

1. About the "Show the 'Send yourself a copy' checkbox" label ...I cannot think of any other place in the admin UI where we offer an option to show/hide form elements. I think that we instead give things titles related to the functionality they enable/disable. I think that the help text of this checkbox is more fitting for it title:

  > Allow users to send themselves a copy of the submitted data.

...and the help text could then be:

  > Exposes a 'Send yourself a copy' checkbox on site-wide and personal contact forms.

2. I feel that using the word "limit" in an element and then using the word "threshold" to explain it is backwards. Think this:

supercalifragilisticexpialidocious
The thing that Mary Poppins used to say.

So, I think that the label of the collapsed fieldset should be "Email send thresholds" ...and the help text could be simplified to something like this:

  > Limit the amount of emails that can be sent from contact forms, from the same IP, within a specific window of time. This helps prevent malicious users from sending multiple emails.

Here's a follow-up PR in case you agree with the above: https://github.com/backdrop/backdrop/pull/2226

![screen shot 2018-06-16 at 7 30 30 am](https://user-images.githubusercontent.com/2423362/41490389-3a940e5e-7137-11e8-88a3-acb3c3b7add3.png)

**Comment by BWPanda** [comment]
@klonos Any review that has a Mary Poppins reference gets my vote 😉😁

**Comment by klonos** [comment]
...moved to #3192
